providers/cloudflare.go

```diff
@@ -0,0 +1,204 @@
+}
+
+func (c *CloudflareHandler) RemoveRecord(record dns.DnsRecord) error {
+	records, err := c.findRecords(record)
+	if err != nil {
+		return err
+	}
+
+	for _, rec := range records {
+		err := c.client.Records.Delete(c.ctx, c.zone.ID, rec.ID)
+		if err != nil {
+			return fmt.Errorf("Cloudflare API call has failed: %v", err)
+		}
+	}
+
+	return nil
+}
+
+func (c *CloudflareHandler) GetRecords() ([]dns.DnsRecord, error) {
+	var records []dns.DnsRecord
+	result, err := c.client.Records.List(c.ctx, c.zone.ID)
+	if err != nil {
+		return records, fmt.Errorf("Cloudflare API call has failed: %v", err)
+	}
+
+	recordMap := map[string]map[string][]string{}
+	recordTTLs := map[string]map[string]int{}
+
+	for _, rec := range result {
+		fqdn := fqdn(rec.Name)
+		recordTTLs[fqdn] = map[string]int{}
+		recordTTLs[fqdn][rec.Type] = rec.TTL
+		recordSet, exists := recordMap[fqdn]
+		if exists {
+			recordSlice, sliceExists := recordSet[rec.Type]
+			if sliceExists {
+				recordSlice = append(recordSlice, rec.Content)
+				recordSet[rec.Type] = recordSlice
+			} else {
+				recordSet[rec.Type] = []string{rec.Content}
+			}
+		} else {
+			recordMap[fqdn] = map[string][]string{}
+			recordMap[fqdn][rec.Type] = []string{rec.Content}
+		}
+	}
+
+	for fqdn, recordSet := range recordMap {
+		for recordType, recordSlice := range recordSet {
+			ttl := recordTTLs[fqdn][recordType]
+			record := dns.DnsRecord{Fqdn: fqdn, Records: recordSlice, Type: recordType, TTL: ttl}
+			records = append(records, record)
+		}
+	}
+	return records, nil
+}
+
+func (c *CloudflareHandler) setZone() error {
+	zones, err := c.client.Zones.List(c.ctx)
+	if err != nil {
+		return fmt.Errorf("Cloudflare API call has failed: %v", err)
+	}
+
+	for _, zone := range zones {
+		if zone.Name == c.root {
+			c.zone = zone
+			break
+		}
+	}
+	if c.zone == nil {
+		return fmt.Errorf("Zone %s does not exist", c.root)
+	}
+
+	return nil
+}
+
+func (c *CloudflareHandler) prepareRecord(record dns.DnsRecord) *cloudflare.Record {
+	name := unFqdn(record.Fqdn)
+	return &cloudflare.Record{
+		Type:   record.Type,
+		Name:   name,
+		TTL:    sanitizeTTL(record.TTL),
+		ZoneID: c.zone.ID,
+	}
+}
+
+func (c *CloudflareHandler) findRecords(record dns.DnsRecord) ([]*cloudflare.Record, error) {
+	var records []*cloudflare.Record
+	result, err := c.client.Records.List(c.ctx, c.zone.ID)
+	if err != nil {
+		return records, fmt.Errorf("Cloudflare API call has failed: %v", err)
+	}
+
+	name := unFqdn(record.Fqdn)
+	for _, rec := range result {
+		if rec.Name == name && rec.Type == record.Type {
+			records = append(records, rec)
+		}
+	}
+
+	return records, nil
+}
+
+// TTL must be between 120 and 86400 seconds
+func sanitizeTTL(ttl int) int {
+	if ttl < 120 {
+		ttl = 120
+	} else if ttl > 86400 {
+		ttl = 86400
+	}
+	return ttl
+}
+
+func fqdn(name string) string {
+	n := len(name)
+	if n == 0 || name[n-1] == '.' {
+		return name
+	}
+	return name + "."
+}
+
+func unFqdn(name string) string {
```

Comment:
@alena1108  Cloudflare does use the FQDN for the record names - they just omit the trailing dot. I am actually in favor of sticking with FQDN as the identifier for `dns.DnsRecord` because it's unambiguous and correspond to how most DNS providers expect as record name (DNSimple is obviously an exception to that).